src/python/turicreate/toolkits/object_detector/_tf_model_architecture.py

```diff
@@ -0,0 +1,390 @@
+# -*- coding: utf-8 -*-
+# Copyright © 2017 Apple Inc. All rights reserved.
+#
+# Use of this source code is governed by a BSD-3-clause license that can
+# be found in the LICENSE.txt file or at https://opensource.org/licenses/BSD-3-Clause
+
+from __future__ import print_function as _
+from __future__ import division as _
+from __future__ import absolute_import as _
+import tensorflow as _tf
+import numpy as _np
+
+
+class ODTensorFlowModel(object):
+
+    def __init__(self, input_image_shape, batch_size, output_size, init_weights, config, is_train=True):
+        _tf.compat.v1.get_default_graph()
+
+        self.config = config
+        self.batch_size = batch_size
+        self.input_image_shape = input_image_shape
+        self.grid_shape = config['grid_shape']
+        self.box_per_cell = config['box_per_cell']
+        self.num_classes = config['num_classes']
+        self.anchors = config['anchors']
+        self.num_anchors = len(self.anchors)
+        self.output_size = output_size
+        self.is_train = is_train  # Set flag for training or val
+
+        # Create placeholders for image and labels
+        self.images = _tf.compat.v1.placeholder(_tf.float32, [self.batch_size, self.input_image_shape[1],
+                                                              self.input_image_shape[2], 3], name='images')
+        self.labels = _tf.compat.v1.placeholder(_tf.float32,
+                                [self.batch_size, self.grid_shape[0], self.grid_shape[1],
+                                 self.box_per_cell, self.num_classes + 5],
+                                name='labels')
+
+        self.init_weights = init_weights
+        self.tf_model = self.tiny_yolo(inputs=self.images, output_size=self.output_size)
+
+        self.global_step = _tf.Variable(0, trainable=False,
+                                        name="global_step")
+
+        self.loss = self.loss_layer(self.tf_model, self.labels)
+
+        self.base_lr = config['learning_rate']
+        self.num_iterations = config['num_iterations']
+
+        self.init_steps = [self.num_iterations // 2, 3 * self.num_iterations // 4, self.num_iterations]
+        self.lrs = [_np.float32(self.base_lr * 10 ** (-i)) for i, step in enumerate(self.init_steps)]
+        self.steps_tf = self.init_steps[:-1]
+
+        self.lr = _tf.compat.v1.train.piecewise_constant(self.global_step, self.steps_tf, self.lrs)
+        # TODO: Evaluate method to update lr in set_learning_rate()
+
+        self.opt = _tf.compat.v1.train.MomentumOptimizer(self.lr, momentum=0.9)
+
+        self.clip_value = self.config.get('gradient_clipping')
+
+        if self.clip_value is not None:
+            grads_and_vars = self.opt.compute_gradients(self.loss)
+            clipped_gradients = [(self.ClipIfNotNone(g, self.clip_value), v) for g, v in grads_and_vars]
+            self.train_op = self.opt.apply_gradients(clipped_gradients, global_step=self.global_step)
```

Comment:
Don't we also need to set `self.train_op` if `self.clip_value` is `None`?

Reply:
I think we can do away with the if condition here. Since we are pre-defining the clip_value (and are aiming to perform gradient clipping where it is helpful), we won't encounter a None value.